src/handlers/psp.ts

```diff
@@ -0,0 +1,436 @@
+      continue;
+    }
+
+    apiValuesBySubscriptionId = {
+      ...apiValuesBySubscriptionId,
+      ...subscriptions.reduce((acc, subscription) => ({ ...acc, [subscription.id]: apiValue }), {}),
+    };
+  }
+
+  return { ...state, apiValuesBySubscriptionId };
+};
+
+const initializeProviders = async (state: State): Promise<State> => {
+  const { config, baseLogOptions } = state;
+
+  const airnodeWallet = ethers.Wallet.fromMnemonic(config.nodeSettings.airnodeWalletMnemonic);
+
+  const evmChains = config.chains.filter((chain: ChainConfig) => chain.type === 'evm');
+  if (isEmpty(evmChains)) {
+    throw new Error('One or more evm compatible chain(s) must be defined in the provided config');
+  }
+  const providerPromises = flatMap(
+    evmChains.map((chain: ChainConfig) =>
+      map(chain.providers, async (chainProvider, providerName) => {
+        const providerLogOptions: node.LogOptions = {
+          ...baseLogOptions,
+          meta: {
+            ...baseLogOptions.meta,
+            chainId: chain.id,
+            providerName,
+          },
+        };
+
+        // Initialize provider specific data
+        const [logs, evmProviderState] = await initializeProvider(chain, chainProvider.url || '');
+        node.logger.logPending(logs, providerLogOptions);
+        if (isNil(evmProviderState)) {
+          node.logger.warn('Failed to initialize provider', providerLogOptions);
+          return null;
+        }
+
+        return {
+          airnodeWallet,
+          chainId: chain.id,
+          providerName,
+          ...evmProviderState,
+        } as ProviderState<EVMProviderState>;
+      })
+    )
+  );
+
+  const providerStates = await Promise.all(providerPromises);
+  const validProviderStates = providerStates.filter((ps) => !isNil(ps)) as ProviderState<EVMProviderState>[];
+
+  return { ...state, providerStates: validProviderStates };
+};
+
+const checkSubscriptionsConditions = async (
+  subscriptions: Id<Subscription>[],
+  apiValuesBySubscriptionId: { [subscriptionId: string]: ethers.BigNumber },
+  contract: ethers.Contract,
+  voidSigner: ethers.VoidSigner,
+  logOptions: node.LogOptions
+) => {
+  const validSubscriptions: CheckedSubscription[] = [];
+  const conditionPromises = subscriptions.map(
+    (subscription) =>
+      checkSubscriptionCondition(subscription, apiValuesBySubscriptionId[subscription.id], contract, voidSigner).then(
+        ([logs, isValid]) => [logs, { subscription, isValid }]
+      ) as Promise<node.LogsData<{ subscription: Id<Subscription>; isValid: boolean }>>
+  );
+  const result = await Promise.all(conditionPromises);
+  result.forEach(([log, data]) => {
+    const subscriptionLogOptions: node.LogOptions = {
+      ...logOptions,
+      additional: {
+        ...logOptions.additional,
+        subscriptionId: data.subscription.id,
+      },
+    };
+    node.logger.logPending(log, subscriptionLogOptions);
+    if (data.isValid) {
+      validSubscriptions.push({
```

Comment:
You can avoid mutation by using `.filter`, eg. :
```
  const conditionPromises = subscriptions.map(
    (subscription) =>
      checkSubscriptionCondition(subscription, apiValuesBySubscriptionId[subscription.id], contract, voidSigner).then(
        ([logs, isValid]) => [logs, {subscription, isValid}]
      ) as Promise<node.LogsData<{ subscription: Id<Subscription>; isValid: boolean }>>
  );
  const result = await Promise.all(conditionPromises);

  return result.filter(([_log, data]) => data.isValid).map(([log, data]) => {
    const subscriptionLogOptions: node.LogOptions = {
      ...logOptions,
      additional: {
        ...logOptions.additional,
        subscriptionId: data.subscription.id,
      },
    };
    node.logger.logPending(log, subscriptionLogOptions);
    return {
      ...data.subscription,
      apiValue: apiValuesBySubscriptionId[data.subscription.id],
    };
  });
```

Reply:
I almost implemented exactly that 🤯 but then I thought that the `forEach` approach was easier to read. I don't feel too strong about this so I might as well use your implementation 💪🏻

Reply:
I'm just trying to keep Andre happy ;)

Reply:
One thing to notice tho is that I want to call `node.logger.logPending()` for all elements in `results` and not just for those whose `data.isValid` is equal to `true`